src/calicost/utils_distribution_fitting.py

```diff
@@ -27,246 +39,378 @@ def convert_params(mean, std):
-    endog : array, (n_samples,)
-        Y values.
-
+    endog : array, (n_samples,)                                                                                                                                                                                                                                                         Y values.
     exog : array, (n_samples, n_features)
         Design matrix.
-
     weights : array, (n_samples,)
         Sample weights.
-
     exposure : array, (n_samples,)
         Multiplication constant outside the exponential term. In scRNA-seq or SRT data, this term is the total UMI count per cell/spot.
     """
-    def __init__(self, endog, exog, weights, exposure, seed=0, **kwds):
-        super(Weighted_NegativeBinomial, self).__init__(endog, exog, **kwds)
+
+    def __init__(self, endog, exog, weights, exposure, tumor_prop=None, seed=0, **kwargs):
+        super().__init__(endog, exog, **kwargs)
+
+        self.tumor_prop = tumor_prop
+
         self.weights = weights
         self.exposure = exposure
+
+        # NB Weight_BetaBinomial does not specify seed
         self.seed = seed
-    #
+
+        # NB __post_init__ validates the expected tumor proportion and handles incrementing instance count.
+        self.__post_init__()
+
+        logger.info(
+            f"Initializing {self.get_ninstance()}th instance of {self.__class__.__name__} model for endog.shape = {endog.shape}."
+        )
+
+    @abstractmethod
     def nloglikeobs(self, params):
-        nb_mean = np.exp(self.exog @ params[:-1]) * self.exposure
-        nb_std = np.sqrt(nb_mean + params[-1] * nb_mean**2)
-        n, p = convert_params(nb_mean, nb_std)
-        llf = scipy.stats.nbinom.logpmf(self.endog, n, p)
-        neg_sum_llf = -llf.dot(self.weights)
-        return neg_sum_llf
-    #
-    def fit(self, start_params=None, maxiter=10000, maxfun=5000, **kwds):
-        self.exog_names.append('alpha')
+        """
+        Negative log-likelihood for the emission model.
+        """
+        pass
+
+    @abstractmethod
+    def get_default_start_params(self):
+        pass
+
+    @abstractmethod
+    def get_ext_param_name(self):
+        """
+        Named parameter in the model.
+        """
+        pass
+
+    @abstractmethod
+    def __post_init__(self):
+        """
+        Validation and customisation for the derived class.
+        E.g. validate the tumor_proportion and increment the instance
+        count of the derived class.
+        """
+        pass
+
+    def get_ninstance(self):
+        """
+        Return the instance count for the given model
+        """
+        return self.ninstance
+
+    def __callback__(self, params):
+        """
+        Define callback for writing parameter chain to file.
+        """
+        print(f"{params} {self.nloglikeobs(params):.6f};")
+
+    def fit(
+        self,
+        start_params=None,
+        maxiter=10_000,
+        maxfun=5_000,
+        write_chain=True,
```

Comment:
Shall we make write_chain to be False by default? I remember it caused problems when I start multiple CalicoST runs in the same directory and all runs tried to write the fitting details to the same file.